File: Final/Code/Project_4/BaiToan5/Ex1_4/generate_complete_bipartite_graph/generate_complete_bipartite_graph.py
class Graph:
    def __init__(self):
        self.vertices = []
        self.edges = {}

    def add_vertex(self):
        v = len(self.vertices)
        self.vertices.append(v)
        self.edges[v] = []
        return v

    def add_edge(self, u, v):
        self.edges[u].append(v)
        self.edges[v].append(u)

    def display(self):
        for v in self.vertices:
            print(f"Đỉnh {v} nối với: {self.edges[v]}")

def generate_complete_bipartite_graph(p, q):
    G = Graph()
    left = [G.add_vertex() for _ in range(p)]
    right = [G.add_vertex() for _ in range(q)]
    for u in left:
        for v in right:
            G.add_edge(u, v)
    return G

p = int(input("Nhập số đỉnh tập trái p: "))
q = int(input("Nhập số đỉnh tập phải q: "))
graph = generate_complete_bipartite_graph(p, q)

graph.display()

File: Final/Code/Project_4/BaiToan5/Ex1_4/generate_complete_bipartite_graph/generate_complete_bipartite_graph.cpp
#include <vector>
#include <iostream>
using namespace std;

class Graph {
private:
    int V;
    vector<vector<int>> adj;

public:
    Graph() : V(0) {}

    int add_vertex() {
        adj.push_back(vector<int>());
        return V++;
    }

    void add_edge(int u, int v) {
        adj[u].push_back(v);
        adj[v].push_back(u);
    }

    void display() {
        for (int i = 0; i < V; ++i) {
            cout << "Đỉnh " << i << " nối với: ";
            for (int j : adj[i])
                cout << j << " ";
            cout << endl;
        }
    }
};

Graph generate_complete_bipartite_graph(int p, int q) {
    Graph G;
    vector<int> left, right;

    for (int i = 0; i < p; ++i)
        left.push_back(G.add_vertex());

    for (int i = 0; i < q; ++i)
        right.push_back(G.add_vertex());

    for (int u : left)
        for (int v : right)
            G.add_edge(u, v);

    return G;
}

int main() {
    int p, q;
    cout << "Nhập số đỉnh (p): ";
    cin >> p;
    cout << "Nhập số đỉnh (q): ";
    cin >> q;

    Graph G = generate_complete_bipartite_graph(p, q);
    G.display();

    return 0;
}